File: src/main/java/com/ilummc/tlib/util/asm/AsmClassLoader.java
package com.ilummc.tlib.util.asm;

public class AsmClassLoader extends ClassLoader {

    private static final class AsmClassLoaderHolder {
        private static AsmClassLoader instance = new AsmClassLoader();
    }

    public static AsmClassLoader getInstance() {
        return AsmClassLoaderHolder.instance;
    }

    private AsmClassLoader() {
        super(AsmClassLoader.class.getClassLoader());
    }

    public static Class<?> createNewClass(String name, byte[] arr) {
        return getInstance().defineClass(name, arr, 0, arr.length, AsmClassLoader.class.getProtectionDomain());
    }

}

File: module/bukkit-nms/src/main/kotlin/taboolib/module/nms/AsmClassLoader.kt
package taboolib.module.nms

import taboolib.common.ClassAppender
import taboolib.common.io.isDebugMode

object AsmClassLoader : ClassLoader(ClassAppender.getClassLoader()) {

    override fun findClass(name: String?): Class<*> {
        try {
            return Class.forName(name, false, AsmClassLoader::class.java.classLoader)
        } catch (e: Throwable) {
            if (isDebugMode) e.printStackTrace()
        }
        return super.findClass(name)
    }

    @JvmStatic
    fun createNewClass(name: String, arr: ByteArray): Class<*> {
        return defineClass(name.replace('/', '.'), arr, 0, arr.size, AsmClassLoader::class.java.protectionDomain)
    }
}